pkg/node/node.go

```diff
@@ -21,3 +38,136 @@ func GetNode(ctx context.Context, nodeName string, kubeClient client.Client) (*v
 	}
 	return node, nil
 }
+
+func EnsureNodePlugins(ctx context.Context, nodeObj *corev1.Node, kubeClient client.Client) error {
+	var foundNvidiaPlugin, foundDADIPlugin bool
+	//does node have vhd installed
+	foundNvidiaPlugin, err := checkAndInstallNvidiaPlugin(ctx, nodeObj, kubeClient)
+	if err != nil {
+		return err
+	}
+	//does node have the custom label for DADI
+	foundDADIPlugin, err = checkAndInstallDADI(ctx, nodeObj, kubeClient)
+	if err != nil {
+		return err
+	}
+	if foundNvidiaPlugin && foundDADIPlugin {
+		// TODO
+	}
+	return nil
+}
+
+func checkAndInstallNvidiaPlugin(ctx context.Context, nodeObj *corev1.Node, kubeClient client.Client) (bool, error) {
+	// check if label accelerator=nvidia exists in the node
+	var foundLabel, foundCapacity bool
+	if nvidiaLabelVal, found := nodeObj.Labels[LabelKeyNvidia]; found {
+		if nvidiaLabelVal == LabelValueNvidia {
+			foundLabel = true
+		} else {
+			nodeObj.Labels = lo.Assign(nodeObj.Labels, map[string]string{LabelKeyCustomGPUProvisioner: GPUString})
+			err := kubeClient.Update(ctx, nodeObj, &client.UpdateOptions{})
+			if err != nil {
+				klog.ErrorS(err, "cannot update node with custom label to enable Nvidia plugin", "node", nodeObj.Name, LabelKeyCustomGPUProvisioner, GPUString)
+				return false, err
+			}
+		}
+	}
+
+	ds := &apps.DaemonSet{}
+	err := retry.OnError(retry.DefaultBackoff, func(err error) bool {
+		return true
+	}, func() error {
+		return kubeClient.Get(ctx, client.ObjectKey{Name: NvidiaDaemonSetName, Namespace: GPUProvisionerNamespace}, ds, &client.GetOptions{})
+	})
+
+	if err != nil {
+		klog.ErrorS(err, "cannot get Nvidia daemonset plugin", "daemonset-name", NvidiaDaemonSetName, "daemonset-namespace", GPUProvisionerNamespace)
+		return false, err
+	}
+
+	if ds.Status.NumberAvailable < 0 {
+		klog.ErrorS(err, "Nvidia daemonset plugin is not running", "daemonset-name", NvidiaDaemonSetName, "daemonset-namespace", GPUProvisionerNamespace)
+		return false, err
+	}
+	podFound, err := checkDaemonSetPodForNode(ctx, NvidiaDaemonSetName, nodeObj.Name, kubeClient)
+	if err != nil {
+		return false, err
+	}
+
+	if podFound {
+		// check Status.Capacity.nvidia.com/gpu has value
+		capacity := nodeObj.Status.Capacity
```

Comment:
For nvidia, you don't need to check ds and pod, just check node status is good enough.

Reply:
Did you make code change here?